diff --git a/dv/prim_alert/prim_alert_sim_cfg.hjson b/dv/prim_alert/prim_alert_sim_cfg.hjson
index 4febee8..e768cd3 100644
--- a/dv/prim_alert/prim_alert_sim_cfg.hjson
+++ b/dv/prim_alert/prim_alert_sim_cfg.hjson
@@ -21,7 +21,7 @@
   testplan: "{proj_root}/hw/ip/prim/dv/prim_alert/data/prim_alert_testplan.hjson"

   // Import additional common sim cfg files.
-  import_cfgs: ["{proj_root}/hw/dv/tools/dvsim/common_sim_cfg.hjson"]
+  import_cfgs: ["{proj_root}/hw/vendor/lowrisc_ip/dv/tools/dvsim/common_sim_cfg.hjson"]

   // Default iterations for all tests - each test entry can override this.
   reseed: 10
diff --git a/dv/prim_esc/prim_esc_sim_cfg.hjson b/dv/prim_esc/prim_esc_sim_cfg.hjson
index 23d297e..a102b9b 100644
--- a/dv/prim_esc/prim_esc_sim_cfg.hjson
+++ b/dv/prim_esc/prim_esc_sim_cfg.hjson
@@ -21,7 +21,7 @@
   testplan: "{proj_root}/hw/ip/prim/dv/prim_esc/data/prim_esc_testplan.hjson"

   // Import additional common sim cfg files.
-  import_cfgs: ["{proj_root}/hw/dv/tools/dvsim/common_sim_cfg.hjson"]
+  import_cfgs: ["{proj_root}/hw/vendor/lowrisc_ip/dv/tools/dvsim/common_sim_cfg.hjson"]

   // Default iterations for all tests - each test entry can override this.
   reseed: 20
diff --git a/dv/prim_lfsr/prim_lfsr_sim_cfg.hjson b/dv/prim_lfsr/prim_lfsr_sim_cfg.hjson
index ea96e83..0a33d8f 100644
--- a/dv/prim_lfsr/prim_lfsr_sim_cfg.hjson
+++ b/dv/prim_lfsr/prim_lfsr_sim_cfg.hjson
@@ -18,7 +18,7 @@
   fusesoc_core: lowrisc:dv:prim_lfsr_sim:0.1

   // Import additional common sim cfg files.
-  import_cfgs: ["{proj_root}/hw/dv/tools/dvsim/common_sim_cfg.hjson"]
+  import_cfgs: ["{proj_root}/hw/vendor/lowrisc_ip/dv/tools/dvsim/common_sim_cfg.hjson"]

   // Default iterations for all tests - each test entry can override this.
   reseed: 1
diff --git a/dv/prim_present/prim_present_sim_cfg.hjson b/dv/prim_present/prim_present_sim_cfg.hjson
index 03c62c6..1fe2df0 100644
--- a/dv/prim_present/prim_present_sim_cfg.hjson
+++ b/dv/prim_present/prim_present_sim_cfg.hjson
@@ -18,7 +18,7 @@
   fusesoc_core: lowrisc:dv:prim_present_sim:0.1

   // Import additional common sim cfg files.
-  import_cfgs: ["{proj_root}/hw/dv/tools/dvsim/common_sim_cfg.hjson"]
+  import_cfgs: ["{proj_root}/hw/vendor/lowrisc_ip/dv/tools/dvsim/common_sim_cfg.hjson"]

   // Default iterations for all tests - each test entry can override this.
   reseed: 1
