commit f331857948a8378334eefd2c407a4e5b86658f7c
Author: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>
Date:   Mon Dec 15 12:24:18 2025 +0000

    REUSE patch

diff --git a/config/gen_from_riscv_config/scripts/libs/utils.py b/config/gen_from_riscv_config/scripts/libs/utils.py
index 93a57452e..4e93487ee 100644
--- a/config/gen_from_riscv_config/scripts/libs/utils.py
+++ b/config/gen_from_riscv_config/scripts/libs/utils.py
@@ -357,10 +357,12 @@ class RstAddressBlock(AddressBlockClass):
         r.directive(
             "..",
             content=[
+                # REUSE-IgnoreStart
                 "Copyright (c) 2024 OpenHW Group",
                 "Copyright (c) 2024 Thales",
                 "SPDX-License-Identifier: Apache-2.0 WITH SHL-2.1",
                 "Author: Abdessamii Oukalrazqou",
+                # REUSE-IgnoreEnd
             ],
         )
         r.title(self.name)  # Use the name of the addressBlock as title
@@ -487,10 +489,12 @@ class InstrstBlock(InstructionBlockClass):
         r.directive(
             "..",
             content=[
+                # REUSE-IgnoreStart
                 "Copyright (c) 2024 OpenHW Group",
                 "Copyright (c) 2024 Thales",
                 "SPDX-License-Identifier: Apache-2.0 WITH SHL-2.1",
                 "Author: Abdessamii Oukalrazqou",
+                # REUSE-IgnoreEnd
             ],
         )
         r.title(self.name)  # Use the name of the addressBlock as title
@@ -567,12 +571,14 @@ class AdocAddressBlock(AddressBlockClass):
         regRV32List = [reg.RV32 for reg in registerlist]
         regRV64List = [reg.RV64 for reg in registerlist]
 
+        # REUSE-IgnoreStart
         r += "////\n"
         r += "  Copyright (c) 2024 OpenHW Group\n"
         r += "  Copyright (c) 2024 Thales\n"
         r += "  SPDX-License-Identifier: Apache-2.0 WITH SHL-2.1\n"
         r += "  Author: Abdessamii Oukalrazqou\n"
         r += "////\n\n"
+        # REUSE-IgnoreEnd
 
         r += "=== %s\n\n"%self.name
         r += "==== Conventions\n\n"
@@ -681,12 +687,14 @@ class InstadocBlock(InstructionBlockClass):
         InstrDescrList = [reg.descr for reg in self.Instructionlist]
         InstrExtList = [reg.Extension_Name for reg in self.Instructionlist]
 
+        # REUSE-IgnoreStart
         r += "////\n"
         r += "  Copyright (c) 2024 OpenHW Group\n"
         r += "  Copyright (c) 2024 Thales\n"
         r += "  SPDX-License-Identifier: Apache-2.0 WITH SHL-2.1\n"
         r += "  Author: Abdessamii Oukalrazqou\n"
         r += "////\n\n"
+        # REUSE-IgnoreEnd
 
         r += "=== %s\n\n"%self.name
         r += "==== Instructions\n\n"
@@ -738,11 +746,13 @@ class InstmdBlock(InstructionBlockClass):
         InstrDescrList = [reg.descr for reg in self.Instructionlist]
         InstrExtList = [reg.Extension_Name for reg in self.Instructionlist]
         licence = [
+            # REUSE-IgnoreStart
             "<!--Copyright (c) 2024 OpenHW Group",
             "Copyright (c) 2024 Thales",
             "SPDX-License-Identifier: Apache-2.0 WITH SHL-2.1 ",
             "Author: Abdessamii Oukalrazqou",
             "-->",
+            # REUSE-IgnoreEnd
         ]
         for l in licence:
             self.mdFile.write(l + "\n")
@@ -847,11 +857,13 @@ class MdAddressBlock(AddressBlockClass):
         ]
         regPrivModeList = [reg.access for reg in registerlist if reg.RV32 | reg.RV64]
         licence = [
+            # REUSE-IgnoreStart
             "<!--Copyright (c) 2024 OpenHW Group",
             "Copyright (c) 2024 Thales",
             "SPDX-License-Identifier: Apache-2.0 WITH SHL-2.1 ",
             "Author: Abdessamii Oukalrazqou",
             "-->",
+            # REUSE-IgnoreEnd
         ]
         for l in licence:
             self.mdFile.write(l + "\n")
diff --git a/docs/scripts/parameters_extractor.py b/docs/scripts/parameters_extractor.py
index e5ef4dd76..cf308771e 100644
--- a/docs/scripts/parameters_extractor.py
+++ b/docs/scripts/parameters_extractor.py
@@ -55,6 +55,7 @@ def writeout_parameter_table(fileout, parameters, module):
 
     with open(fileout, "w") as fout:
         fout.write("..\n")
+        # REUSE-IgnoreStart
         fout.write("   Copyright 2024 Thales DIS France SAS\n")
         fout.write(
             '   Licensed under the Solderpad Hardware License, Version 2.1 (the "License");\n'
@@ -67,6 +68,7 @@ def writeout_parameter_table(fileout, parameters, module):
             "   You may obtain a copy of the License at https://solderpad.org/licenses/\n\n"
         )
         fout.write("   Original Author: Jean-Roch COULON - Thales\n\n")
+        # REUSE-IgnoreEnd
         fout.write(f".. _{module}_PARAMETERS:\n\n")
         fout.write(f".. list-table:: {module} parameter configuration\n")
         fout.write("   :header-rows: 1\n")
@@ -84,6 +86,7 @@ def writeout_parameter_table_adoc(fileout, parameters, module):
 
     with open(fileout, "w") as fout:
         fout.write("////\n")
+        # REUSE-IgnoreStart
         fout.write("   Copyright 2024 Thales DIS France SAS\n")
         fout.write(
             '   Licensed under the Solderpad Hardware License, Version 2.1 (the "License");\n'
@@ -96,6 +99,7 @@ def writeout_parameter_table_adoc(fileout, parameters, module):
             "   You may obtain a copy of the License at https://solderpad.org/licenses/\n\n"
         )
         fout.write("   Original Author: Jean-Roch COULON - Thales\n")
+        # REUSE-IgnoreEnd
         fout.write("////\n\n")
 
         fout.write(f"[[{module}_PARAMETERS]]\n\n")
