diff --git a/common.core b/common.core
index aa2ac3c..4cf2cb6 100644
--- a/common.core
+++ b/common.core
@@ -20,17 +20,12 @@ filesets:
       - tools/veriblelint/lowrisc-styleguide.rules.verible_lint: {file_type: veribleLintRules}
       - tools/veriblelint/common.vbl: {file_type: veribleLintWaiver}
 
-  files_check_tool_requirements:
-    depend:
-     - lowrisc:tool:check_tool_requirements
-
 targets:
   default: &default_target
     filesets:
       - tool_verilator   ? (files_verilator)
       - tool_ascentlint  ? (files_ascentlint)
       - tool_veriblelint ? (files_veriblelint)
-      - files_check_tool_requirements
     tools:
       ascentlint:
         ascentlint_options:
